main.tf

```diff
@@ -29,37 +55,53 @@ resource "aws_launch_configuration" "this" {
-  vpc_zone_identifier  = ["${var.vpc_zone_identifier}"]
-  max_size             = "${var.max_size}"
-  min_size             = "${var.min_size}"
-  desired_capacity     = "${var.desired_capacity}"
-
-  load_balancers            = ["${var.load_balancers}"]
-  health_check_grace_period = "${var.health_check_grace_period}"
-  health_check_type         = "${var.health_check_type}"
-
-  min_elb_capacity          = "${var.min_elb_capacity}"
-  wait_for_elb_capacity     = "${var.wait_for_elb_capacity}"
-  target_group_arns         = ["${var.target_group_arns}"]
-  default_cooldown          = "${var.default_cooldown}"
-  force_delete              = "${var.force_delete}"
-  termination_policies      = "${var.termination_policies}"
-  suspended_processes       = "${var.suspended_processes}"
-  placement_group           = "${var.placement_group}"
-  enabled_metrics           = ["${var.enabled_metrics}"]
-  metrics_granularity       = "${var.metrics_granularity}"
-  wait_for_capacity_timeout = "${var.wait_for_capacity_timeout}"
-  protect_from_scale_in     = "${var.protect_from_scale_in}"
-
-  tags = ["${concat(
-      list(map("key", "Name", "value", var.name, "propagate_at_launch", true)),
+  count = var.create_asg && false == var.create_asg_with_initial_lifecycle_hook ? 1 : 0
+
+  name_prefix = "${join(
+    "-",
+    compact(
+      [
+        coalesce(var.asg_name, var.name),
+        var.recreate_asg_when_lc_changes ? element(concat(random_pet.asg_name.*.id, [""]), 0) : "",
+      ],
+    ),
+  )}-"
+  launch_configuration = var.create_lc ? element(concat(aws_launch_configuration.this.*.name, [""]), 0) : var.launch_configuration
+  vpc_zone_identifier  = var.vpc_zone_identifier
+  max_size             = var.max_size
+  min_size             = var.min_size
+  desired_capacity     = var.desired_capacity
+
+  load_balancers            = var.load_balancers
+  health_check_grace_period = var.health_check_grace_period
+  health_check_type         = var.health_check_type
+
+  min_elb_capacity          = var.min_elb_capacity
+  wait_for_elb_capacity     = var.wait_for_elb_capacity
+  target_group_arns         = var.target_group_arns
+  default_cooldown          = var.default_cooldown
+  force_delete              = var.force_delete
+  termination_policies      = var.termination_policies
+  suspended_processes       = var.suspended_processes
+  placement_group           = var.placement_group
+  enabled_metrics           = var.enabled_metrics
+  metrics_granularity       = var.metrics_granularity
+  wait_for_capacity_timeout = var.wait_for_capacity_timeout
+  protect_from_scale_in     = var.protect_from_scale_in
+
+  tags = [
+    concat(
+      [
+        {
+          "key"                 = "Name"
+          "value"               = var.name
+          "propagate_at_launch" = true
+        },
+      ],
       var.tags,
-      local.tags_asg_format
-   )}"]
+      local.tags_asg_format,
+    ),
+  ]
```

Comment:
And here - possible should be replaced brackets with parentheses in tags =[]

Reply:
Did this, so now its `tags = ()`

Reply:
@drpebcak Regarding this nested block, I am not sure that this is works in this way I was wrote.
It looks like it migh be requires to use `dynamic` block instead of `concat`.
Sorry for confusing you- I habe not a big expiriense with nested block expressions

Reply:
Just noticed there is another instance of tags just like this one that should match. I think concat is fine because we are just taking 3 lists and combining them. We probably dont need to wrap . the concat in anything.

Reply:
regarding this null resourse: with the curent code terraform plan returns me the following:
Error: Incorrect attribute value type

  on ../../tf_modules/aws/asg/main.tf line 86, in resource "aws_autoscaling_group" "this":
  86:   tags = concat(

Inappropriate value for attribute "tags": element 4: map of string required.

In case if I comment local.tags_asg_format in module - than everything okay.

Reply:
fixed on my side by removing brackets in tags_asg_format = null_resource.tags_as_list_of_maps.*.triggers but I am not sure that this is correct

Reply:
Ah, yes of course. `null_resource.tags_as_list_of_maps.*.triggers` is already an array. I've added that fix as well.